Adding lichess data tests

janakerman opened the pull request:
Resolves #24.

This adds sort of unit, sort of integration tests (really want opinions here!).

The tests hit the newly created LichessDataService and use the HttpUtility to stub out HTTP responses for given end points. This is testing the parsing and fetching logic of Lichess entities.

Another example of tests that will be added into this file are the tests that will come in to ensure that the LichessDataService is saving the games to our repository as they come in, etc.

janakerman committed:
Upgrade Groovy version to 2.4.10

janakerman committed:
Upgrade spock versions & adds spring spock

janakerman committed:
Inlines cahe name

For some reason Groovy is complaining again about only using
inlined strings withing annotations. This seems to be tempermental..

janakerman committed:
Adds LichessDataService

EntityService now just forwards through to the data service
until we get our DB connected up. The Lichess data service
will eventually run from a scheduler.

janakerman committed:
Adds LichessDataService unit tests

Tests for games will follow.

janakerman committed:
Use JSON files for unit tests instead of generating in code

This adds a bunch of mock Lichess responses, 2 pages for 3 users
and two pages of team member data. The page sizes in the tests are
set intentionally small so that we get paged results for the small team we
have.

The page sizes have been pulled out into application.properties, this allows
us to inject them in our test env, but is nicer anyway as a bonus. The page sizes
are also now a responsibility of the LichessDataService, and not the LichessApi
which now takes page size as a parameter.

bjedrzejewski commented on src/main/java/com/chessmates/service/EntityServiceImpl.groovy:
```diff
@@ -2,130 +2,35 @@ package com.chessmates.service
-        resultSet.get()
-    }
+    // TODO: Replace with fetch from DB.
+    List<Player> getPlayers() { lichessDataService.getPlayers(null) }
```
How about we make a version of this method that takes no parameter rather than passing null

bjedrzejewski commented on src/main/java/com/chessmates/service/EntityServiceImpl.groovy:
```diff
@@ -2,130 +2,35 @@ package com.chessmates.service
-
-        def playerCombinations = uniqueCombinations(players)
+    // TODO: Replace with fetch from DB.
+    List<Game> getGames() { lichessDataService.getGames(getPlayers(), null) }
```
Maybe we can make a version that does not take a null, but just takes the players

bjedrzejewski commented on src/main/java/com/chessmates/utility/LichessApiImpl.groovy:
```diff
@@ -91,19 +91,33 @@ class LichessApiImpl implements LichessApi {
+        final playerId = (String)playerObject?.id
+
+        if (!playerId) {
+            logger.warn PARSE_ERROR
```
Can you maybe add some details about what failed parsing?

janakerman replied:
Added into the exception itself

bjedrzejewski commented on src/main/java/com/chessmates/utility/LichessApiImpl.groovy:
```diff
@@ -78,6 +77,7 @@ class LichessApiImpl implements LichessApi {
 
         def pageResults = resultObjects.stream()
                 .map(parse)
+                .filter {item -> item != null }
```
Maybe you can handle it with exceptions rather than null checking.

bjedrzejewski replied:
The other option is to use Optional of Game https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

janakerman replied:
Let's go with exceptions here - love the idea of using optionals to make nulls explicit

bjedrzejewski commented on src/test/java/com/chessmates/service/LichessDataServiceImplTest.groovy:
```diff
@@ -0,0 +1,246 @@
+    @TestConfiguration
+    private static class MockConfig {
+
+        final detatchedMockFactory = new DetachedMockFactory()
```
Spelling mistake in detatchedMockFactory

bjedrzejewski reviewed (changes requested):
Minor tweaks

janakerman committed:
Fixes issue from PR feedback

- Spelling errors
- Parses throw exception, users handle that
- Adds parameter less methods to LichessDataService API.
